### PR TITLE
[corrosion] Add the Corrosion CMake Tool

### PR DESCRIPTION
**Pull request by gauntl3t12**
- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [X] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [X] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [X] The versioning scheme in `vcpkg.json` matches what upstream says.
- [X] The license declaration in `vcpkg.json` matches what upstream says.
- [X] The installed as the "copyright" file matches what upstream says.
- [X] The source code of the component installed comes from an authoritative source.
- [X] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by bconn98** [comment]
It wasn't originally clear to me that you want each version in a separate MR (if I'm reading the MR template correctly). I can back out 0.4.1+ and submit the other 7 as separate MRs.

**Comment by bconn98** [comment]
@microsoft-github-policy-service agree

**Comment by bconn98** [comment]
Closing to relaunch as individual MRs with fixes applied for platforms other than linux

**Comment by dg0yt** [comment]
IMO it is really enough to add only 0.4.7 from the 0.4 series. By common conventions, these are patch releases.

**Comment by bconn98** [comment]
Reopening as only submitting 1 version per @dg0yt suggestion

File: ports/corrosion/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    SHA512 274baca57f7d599b304b75a73067fae9eb488eec10925fade7e195d494a192760b116a3bdf289e0cb7c291b29684909d5fd1c9404c6d37203c883cd511849bbb
+    HEAD_REF master
+    PATCHES
+        "vcpkg-install.patch"
+)
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+        -DCORROSION_VCPKG_INSTALL=ON
```
Comment (dg0yt) [inline ports/corrosion/portfile.cmake:16]: IMO there shouldn't be a "vcpkg" option. Patch it to do the right thing in vcpkg, or patch it in a generic way for upstreaming.

Comment (bconn98) [reply ports/corrosion/portfile.cmake:16]: Agreed, I'll be putting a PR out against the corrosion repo. The patch when applied, is just following the comment supplied within their main CMakeLists.

Comment (bconn98) [reply ports/corrosion/portfile.cmake:16]: https://github.com/corrosion-rs/corrosion/pull/485

Comment (bconn98) [reply ports/corrosion/portfile.cmake:16]: @dg0yt I'm open to resubmitting with the same suggested changes/naming if that's preferred.

File: ports/corrosion/usage [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    corrosion_add_target_local_rustflags(target_name rustc_flag ...)
+    corrosion_set_hostbuild(<target_name>)
+    corrosion_set_features(<target_name> [ALL_FEATURES <Bool>] ...)
+    corrosion_set_cargo_flags(<target_name> <flag1> ...])
+    corrosion_set_linker(target_name linker)
+    corrosion_cbindgen(TARGET <imported_target_name> CARGO_PACKAGE ...)
+    corrosion_add_cxxbridge(cxx_target CRATE <imported_tgt_name> FILES lib.rs)
+
+See https://corrosion-rs.github.io/corrosion/introduction.html
+
```
Comment (MonicaLiu0311) [inline ports/corrosion/usage:16]: Here we only need `find_package(Corrosion CONFIG REQUIRED)` and no need for `target_link_libraries(...)`?

Sorry I don't know much about this tool, but just after `find_package()` I get the following error:
```
1> [CMake] CMake Error at G:/corrosion/installed/x64-windows/share/cmake/FindRust.cmake:23 (message):
1> [CMake]   `rustc` not found in PATH or `/.cargo/bin`.
1> [CMake] 
1> [CMake]   Hint: Check if `rustc` is in PATH or manually specify the location by
1> [CMake]   setting `Rust_COMPILER` to the path to `rustc`.
1> [CMake] Call Stack (most recent call first):
1> [CMake]   G:/corrosion/installed/x64-windows/share/cmake/FindRust.cmake:228 (_findrust_failed)
1> [CMake]   G:/corrosion/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
1> [CMake]   G:/corrosion/installed/x64-windows/share/cmake/Corrosion.cmake:63 (find_package)
1> [CMake]   G:/corrosion/installed/x64-windows/share/corrosion/CorrosionConfig.cmake:43 (include)
1> [CMake]   G:/corrosion/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
1> [CMake]   CMakeUsage/CMakeLists.txt:15 (find_package)
1> [CMake] -- Configuring incomplete, errors occurred!
```
Can you guide me on how to use it, or provide me with a test case?

Comment (MonicaLiu0311) [reply ports/corrosion/usage:16]: Following [an example project](https://github.com/bconn98/logger) you provided, my test passed:

```
G:\logger\build>cmake ..
-- Building for: Visual Studio 17 2022
-- Running vcpkg install
Fetching registry information from https://github.com/bconn98/vcpkg (HEAD)...
Detecting compiler hash for triplet x64-windows...
The following packages will be built and installed:
    corrosion:x64-windows@0.4.7 -- C:\Users\monica\AppData\Local\vcpkg\registries\git-trees\a2a4a7f65d35df93a2dff5e3359351f7d8474314
  * vcpkg-cmake:x64-windows@2023-05-04 -- C:\Users\monica\AppData\Local\vcpkg\registries\git-trees\88a7058fc7fa73a9c4c99cfcae9d79e2abf87a5a
  * vcpkg-cmake-config:x64-windows@2022-02-06#1 -- C:\Users\monica\AppData\Local\vcpkg\registries\git-trees\8d54cc4f487d51b655abec5f9c9c3f86ca83311f
Additional packages (*) will be modified to complete this operation.
Restored 3 package(s) from C:\Users\monica\AppData\Local\vcpkg\archives in 84.9 ms. Use --debug to see more details.
Installing 1/3 vcpkg-cmake-config:x64-windows@2022-02-06#1...
Elapsed time to handle vcpkg-cmake-config:x64-windows: 5.11 ms
vcpkg-cmake-config:x64-windows package ABI: 69447b17296abbb28673fffa972ded1496288e510092d15899df8d3236f5e14a
Installing 2/3 vcpkg-cmake:x64-windows@2023-05-04...
Elapsed time to handle vcpkg-cmake:x64-windows: 11.3 ms
vcpkg-cmake:x64-windows package ABI: cddc3abc622c04922f3858562a12a34f26fe43c5460b691e0ab690bd7295c54b
Installing 3/3 corrosion:x64-windows@0.4.7...
Elapsed time to handle corrosion:x64-windows: 18.3 ms
corrosion:x64-windows package ABI: c4d515768e318a67ac7c68d9172b325032b7e5ebad7b9dcdeaf69c7d313bc374
Total install time: 34.8 ms
The package corrosion provides additional cmake functions:

    find_package(Corrosion CONFIG REQUIRED)
    corrosion_import_crate(MANIFEST_PATH <path/to/cargo.toml> ...)
    corrosion_set_env_vars(<target_name> <key1=value1> [... <keyN=valueN>])
    corrosion_add_target_rustflags(<target_name> <rustflag> [... <rustflagN>])
    corrosion_add_target_local_rustflags(target_name rustc_flag ...)
    corrosion_set_hostbuild(<target_name>)
    corrosion_set_features(<target_name> [ALL_FEATURES <Bool>] ...)
    corrosion_set_cargo_flags(<target_name> <flag1> ...])
    corrosion_set_linker(target_name linker)
    corrosion_cbindgen(TARGET <imported_target_name> CARGO_PACKAGE ...)
    corrosion_add_cxxbridge(cxx_target CRATE <imported_tgt_name> FILES lib.rs)

See https://corrosion-rs.github.io/corrosion/introduction.html


-- Running vcpkg install - done
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.19045.
-- The CXX compiler identification is MSVC 19.38.33134.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Enterprise/VC/Tools/MSVC/14.38.33130/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Rust Toolchain: stable-x86_64-pc-windows-msvc
-- Rust Target: x86_64-pc-windows-msvc
-- Determining required link libraries for target x86_64-pc-windows-msvc
-- Required static libs for target x86_64-pc-windows-msvc: kernel32;advapi32;bcrypt;kernel32;ntdll;userenv;ws2_32;kernel32;ws2_32;kernel32;ntdll;kernel32;$<$<CONFIG:Debug>:msvcrtd>
-- Found Rust: C:/Users/monica/.rustup/toolchains/stable-x86_64-pc-windows-msvc/bin/rustc.exe (found version "1.75.0")
-- Using Corrosion as a subdirectory
-- Configuring done (12.6s)
-- Generating done (0.1s)
-- Build files have been written to: G:/logger/build

G:\logger\build>
G:\logger\build>
G:\logger\build>cmake --build .
MSBuild version 17.8.3+195e7f5a3 for .NET Framework

  1>Checking Build System
  Building Custom Rule G:/logger/CMakeLists.txt
     Compiling proc-macro2 v1.0.78
     Compiling unicode-ident v1.0.12
     Compiling autocfg v1.1.0
     Compiling serde v1.0.196
     Compiling cc v1.0.83
     Compiling windows_x86_64_msvc v0.48.5
     Compiling windows_x86_64_msvc v0.52.0
     Compiling parking_lot_core v0.9.9
     Compiling winapi v0.3.9
     Compiling syn v1.0.109
     Compiling num-traits v0.2.17
     Compiling indexmap v1.9.3
     Compiling lock_api v0.4.11
     Compiling link-cplusplus v1.0.9
     Compiling windows-targets v0.48.5
     Compiling cxxbridge-flags v1.0.115
     Compiling quote v1.0.35
     Compiling serde_json v1.0.113
     Compiling cfg-if v1.0.0
     Compiling smallvec v1.13.1
     Compiling linked-hash-map v0.5.6
     Compiling hashbrown v0.12.3
     Compiling destructure_traitobject v0.2.0
     Compiling ryu v1.0.16
     Compiling syn v2.0.48
     Compiling thiserror v1.0.56
     Compiling anyhow v1.0.79
     Compiling scopeguard v1.2.0
     Compiling ordered-float v2.10.1
     Compiling yaml-rust v0.4.5
     Compiling unsafe-any-ors v1.0.0
     Compiling cxx v1.0.115
     Compiling windows-targets v0.52.0
     Compiling itoa v1.0.10
     Compiling chrono v0.4.33
     Compiling typemap-ors v1.0.0
     Compiling parking_lot v0.12.1
     Compiling humantime v2.1.0
     Compiling thread-id v4.2.1
     Compiling log-mdc v0.1.0
     Compiling arc-swap v1.6.0
     Compiling fnv v1.0.7
     Compiling serde_derive v1.0.196
     Compiling thiserror-impl v1.0.56
     Compiling cxxbridge-macro v1.0.115
     Compiling derivative v2.2.0
     Compiling serde-value v0.7.0
     Compiling log v0.4.20
     Compiling serde_yaml v0.8.26
     Compiling log4rs v1.2.0
     Compiling log-layer v0.1.0 (G:\logger\log-layer)
      Finished dev [unoptimized + debuginfo] target(s) in 36.85s
  Building Custom Rule G:/logger/CMakeLists.txt
  Copying byproducts `log_layer.lib` to G:/logger/build
  Building Custom Rule G:/logger/CMakeLists.txt
  Building Custom Rule G:/logger/CMakeLists.txt
  Generating cxx bindings for crate log-layer
  Generating rust/cxx.h header
  Building Custom Rule G:/logger/CMakeLists.txt
  lib.cpp
  cxx_bridge.vcxproj -> G:\logger\build\Debug\cxx_bridge.lib
  Building Custom Rule G:/logger/CMakeLists.txt
  main.cpp
  mylib.cpp
  Generating Code...
  log-example.vcxproj -> G:\logger\build\Debug\log-example.exe
  Building Custom Rule G:/logger/CMakeLists.txt
```


<details>
<summary><b>CMakeLists.txt</b></summary>
<pre>

cmake_minimum_required(VERSION 3.23)

set(CMAKE_TOOLCHAIN_FILE G:/corrosion/scripts/buildsystems/vcpkg.cmake)
set(CMAKE_EXPORT_COMPILE_COMMANDS true)
set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_FLAGS " /EHsc /RTC1 ")
set(CMAKE_MSVC_RUNTIME_LIBRARY MultiThreadedDLL)

project(log-example LANGUAGES CXX VERSION 0.0.1)

####################### Corrosion Linking ############################
find_package(Corrosion CONFIG REQUIRED)

if (WIN32)
    set(CXXBRIDGE_HOME "C:/Users/monica/.cargo/bin/")
else ()
    set(CXXBRIDGE_HOME "$ENV{HOME}/.cargo/bin/")
endif()

find_program(CXXBRIDGE cxxbridge PATHS ${CXXBRIDGE_HOME})
if (CXXBRIDGE STREQUAL "CXXBRIDGE-NOTFOUND")
    message("Could not find cxxbridge, trying to install with `cargo install cxxbridge-cmd'")
    find_program(CARGO cargo PATHS ${CXXBRIDGE_HOME})
    if (CARGO STREQUAL "CARGO-NOTFOUND")
        message(FATAL_ERROR "Requires cargo available in path, install via rustup https://rustup.rs/")
    endif()
    
    # 在 Windows 上，可能需要使用 cmd.exe 来执行 cargo 命令
    if (WIN32)
        execute_process(COMMAND cmd /c "${CARGO} install cxxbridge-cmd")
    else ()
        execute_process(COMMAND ${CARGO} install cxxbridge-cmd)
    endif()
    
    find_program(CXXBRIDGE cxxbridge PATHS ${CXXBRIDGE_HOME})
endif()

corrosion_import_crate(MANIFEST_PATH log-layer/Cargo.toml)
corrosion_add_cxxbridge(cxx_bridge CRATE log-layer MANIFEST_PATH log-layer FILES lib.rs)

####################### Corrosion Linking ############################

add_executable(${PROJECT_NAME} src/main.cpp src/test/mylib.cpp)
target_include_directories(${PROJECT_NAME} PRIVATE "inc" "inc/test")

target_link_options(${PROJECT_NAME} PRIVATE "/NODEFAULTLIB:MSVCRT")

target_link_libraries(${PROJECT_NAME} PRIVATE cxx_bridge)

</pre></details>

Comment (MonicaLiu0311) [reply ports/corrosion/usage:16]: However, Rust tools have not been added to vcpkg in any form yet. Related work: #30251, #31748.

**Review by MonicaLiu0311 — APPROVED** [review]

**Comment by MonicaLiu0311** [comment]
Mark `info:reviewed` for further review.

Note: It has a dependency on both the Rust compiler (cargo) and a C++ compiler.

**Comment by MonicaLiu0311** [comment]
Mark `info:reviewed` for further review.

Note: It has a dependency on both the Rust compiler (cargo) and a C++ compiler.

**Comment by jschwe** [comment]
> Note: It has a dependency on both the Rust compiler (cargo) and a C++ compiler.

Corrosion maintainer here. Just to clarify, a C++ compiler is not required from Corrosions side. If you want to integrate Rust code into a C++ project, then of course a C++ compiler is required to compile and link the C++ code. But it's perfectly possible to use Corrosion with C + Rust projects, in which case a C-compiler is sufficient, or even just use Corrosion to only build rust code without building or linking any C or C++ code.

**Comment by data-queue** [comment]
Vcpkg generally does not accept a port with only CMake helpers unless a specific C++ library requires it to build. Can you give an example? Also, I'm not exactly sure what benefit this port provides over the upstream recommended way of acquiring these helpers with `FetchContent`.

**Comment by bconn98** [comment]
The test project I linked too was used a proof of concept to convenience my team of porting common utility code from C++ to open source rust projects. The projects we've updated are all behind proxies and cannot be shared on public GitHub.  
  
The issue with FetchContent is that Vcpkg disables it by default. As my team develops code released on internal private Vcpkg registries, our port now need to go enable FetchContent again. That's a poor solution and requires us to maintain two different forms of dependency version control. Vcpkg shouldn't really make the statement that everything should come from Vcpkg, but also things should come from features that we disable by default. It wasn't clear or well documented that you even disable FetchContent. It took me 2-3 hours to figure out and resolve the issue.

**Comment by bconn98** [comment]
Also, [cmakerc](https://github.com/microsoft/vcpkg/tree/master/ports/cmakerc) is exactly that a cmake helper.

**Comment by vicroms** [comment]
Hi @bconn98
 
The vcpkg team had a discussion regarding this PR and interactions between C++ and Rust projects. Here’s a summary of our thoughts:
 
We believe that the integration of Rust dependencies in C++ projects is becoming increasingly important and we want vcpkg to provide the best experience possible for C++ developers.
 
However, we are wary about two potential issues:
 
* **Unresolved technical issues** that may arise from this integration. For example, what happens when a cargo package requires vcpkg-installed dependencies?
* The **implications for acquiring Rust dependencies within the vcpkg ecosystem** that this PR could create.
 
Given that you are already using private registries, we believe that keeping this port as part of your team’s private registry is the best way forward. It also lets you make changes as needed without concern of breaking external users.
 
Please let us know if you have any further questions or concerns.

**Comment by bconn98** [comment]
Unfortunate but I understand.

**Comment by vicroms** [comment]
Closed by reasoning explained here: https://github.com/microsoft/vcpkg/pull/36369#issuecomment-1966055372

We may revise our stance in the future once the team has spent time designing a good Rust integration experience.

**Comment by ArcticLampyrid** [comment]
> Vcpkg generally does not accept a port with only CMake helpers unless a specific C++ library requires it to build. Can you give an example?

eg. [Slint](https://slint.dev/), a declarative GUI developed by some former Qt developers.

**Comment by gracicot** [comment]
Note that cargo already support installing vcpkg dependencies: https://crates.io/crates/cargo-vcpkg. If vcpkg supports the other way around, it would be great if it would play nice with scenarios like `vcpkg -> cargo -> vcpkg` or `cargo -> vcpkg -> cargo`

**Pull request closed without merging**